### PR TITLE
Add FSharp incremental backend lexer

### PR DESCRIPTION
**Pull request by misonijnik**

### CODE REVIEW

File: ReSharper.FSharp/src/FSharp.Psi/src/Parsing/FSharpTokenInfoEx.cs [hunk 1]
```diff
@@ -39,19 +39,16 @@ public static TokenNodeType GetTokenType(this FSharpToken token)
 
         case FSharpTokenCharKind.String:
           if (tokenTag == Parser.token.Tags.CHAR)
-            return FSharpTokenType.CHAR;
+            return FSharpTokenType.CHARACTER_LITERAL;
```
[the remark this answers is not in the document]
Comment (misonijnik) [reply ReSharper.FSharp/src/FSharp.Psi/src/Parsing/FSharpTokenInfoEx.cs]: No, I'll delete it.

File: ReSharper.FSharp/src/FSharp.Psi.Features/src/CodeCompletion/FSharpKeywordsProvider.fs [hunk 2]
```diff
@@ -59,7 +59,6 @@ type FSharpKeywordsProvider() =
         | _, tokenBeforeType ->
 
         if tokenBeforeType == FSharpTokenType.LINE_COMMENT ||
-           tokenBeforeType == FSharpTokenType.DEAD_CODE ||
```
[the remark this answers is not in the document]
Comment (misonijnik) [reply ReSharper.FSharp/src/FSharp.Psi.Features/src/CodeCompletion/FSharpKeywordsProvider.fs]: Well, I'll comment on the code associated with the preprocessor.

File: ReSharper.FSharp/src/FSharp.Psi.Features/src/Parsing/FSharpParser.fs [hunk 3]
```diff
@@ -1,16 +1,22 @@
...
+open JetBrains.ReSharper.Plugins.FSharp.Psi.Features.Parsing.Lexing
 open JetBrains.ReSharper.Plugins.FSharp.Psi.Parsing
 open JetBrains.ReSharper.Plugins.FSharp.Psi.Impl.Tree
+open JetBrains.ReSharper.Plugins.FSharp.Psi.Parsing.Lexing
 open JetBrains.Util
 open Microsoft.FSharp.Compiler.Ast
 open Microsoft.FSharp.Compiler.SourceCodeServices
 
 type internal FSharpParser(file: IPsiSourceFile, checkerService: FSharpCheckerService, logger: ILogger) =
+    let factory = FSharpPreprocessedLexerFactory(checkerService.GetDefines(file)) :> ILexerFactory
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi.Features/src/Parsing/FSharpParser.fs]: Why is it moved out of `ParseFile`?

File: ReSharper.FSharp/src/FSharp.Psi.Features/src/StringLiteralsUtil.fs [hunk 4]
```diff
@@ -1,4 +1,7 @@
 [<AutoOpen>]
+#if KEK
+let kek = "dwqdqw"
+#endif
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi.Features/src/StringLiteralsUtil.fs]: Well, wat? :)

File: ReSharper.FSharp/src/Services.FSharp/src/CodeCompletion/FSharpCodeCompletionContextProvider.cs [hunk 5]
```diff
@@ -38,7 +38,6 @@ public override ISpecificCodeCompletionContext GetCompletionContext(CodeCompleti
 
       if (tokenBeforeType == FSharpTokenType.LINE_COMMENT ||
           tokenBeforeType == FSharpTokenType.DEAD_CODE || tokenType == FSharpTokenType.DEAD_CODE ||
-
```
Comment (auduchinok) [inline ReSharper.FSharp/src/Services.FSharp/src/CodeCompletion/FSharpCodeCompletionContextProvider.cs]: ?

File: ReSharper.FSharp/src/FSharp.Psi/src/Parsing/Lexing/FSharpLexerGenerated.Supplement.cs [hunk 6]
```diff
@@ -0,0 +1,401 @@
+using System;
+using System.Diagnostics;
+using JetBrains.Annotations;
+using JetBrains.ReSharper.Psi.Parsing;
+using JetBrains.Text;
+using JetBrains.Util;
+using JetBrains.Util.DataStructures;
+using JetBrains.Util.Logging;
+using Xunit;
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi/src/Parsing/Lexing/FSharpLexerGenerated.Supplement.cs]: Is it really needed here?

File: ReSharper.FSharp/src/FSharp.Psi/src/Parsing/FSharpPreprocessedLexer.cs [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+        tokenType = myLexer.TokenType;
+        myQueue = myQueue.Enqueue(new Tuple<FSharpLexerState, string>(
+          myLexer.CurrentPosition,
+          myLexer.GetCurrTokenText())
+        );
+        myLexer.Advance();
+      } while (!(tokenType == FSharpTokenType.NEW_LINE || tokenType == null));
+      var list = ListModule.OfSeq(myQueue
+        .Where(x => !Skip(x.Item1.currTokenType))
+        .Select(x => new Tuple<TokenNodeType, string>(x.Item1.currTokenType, x.Item2)));
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi/src/Parsing/FSharpPreprocessedLexer.cs]: I'd try avoid using Linq and tuples here, it looks like too many allocations for a preprocessor.

File: ReSharper.FSharp/src/FSharp.Psi.Features/src/StringLiteralsUtil.fs (hunk 4 above)
Comment (misonijnik) [reply ReSharper.FSharp/src/FSharp.Psi.Features/src/StringLiteralsUtil.fs]: Oops

**Comment by auduchinok** [comment]
Please also check https://youtrack.jetbrains.com/issue/RIDER-20603.

File: ReSharper.FSharp/src/FSharp.Psi/src/Parsing/FSharpPreprocessedLexer.cs [hunk 8]
```diff
@@ -0,0 +1,320 @@
...
+namespace JetBrains.ReSharper.Plugins.FSharp.Psi.Parsing
+{
+  public class FSharpPreprocessedLexer : ILexer<FSharpLexerState>
+  {
+    private TokenNodeType currTokenType;
+    private readonly FSharpLexer myLexer;
+    private readonly FSharpPreprocessor myPreprocessor;
+    private readonly HashSet<string> myDefinedConstants;
+    private PreprocessorState myState = new PreprocessorState();
+    private static readonly NodeTypeSet myDirective = new NodeTypeSet(
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi/src/Parsing/FSharpPreprocessedLexer.cs]: Where it is used at?

File: ReSharper.FSharp/src/FSharp.Psi/src/Parsing/FSharpPreprocessedLexer.cs [hunk 9]
```diff
@@ -0,0 +1,320 @@
...
+              myElseState = PreprocessorElseState.Error;
+              break;
+          }
+          return myElseState;
+        }
+      }
+
+      private Stack<PreprocessorBlockState> myStack = new Stack<PreprocessorBlockState>();
+      private Queue<FSharpLexerState> myQueue = new Queue<FSharpLexerState>();
+      public FSharpLexerState? DumpedState { get; private set; }
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi/src/Parsing/FSharpPreprocessedLexer.cs]: Is this property ever accessed?

File: ReSharper.FSharp/src/FSharp.Psi/src/Parsing/FSharpPreprocessedLexer.cs [hunk 10]
```diff
@@ -0,0 +1,320 @@
...
+
+namespace JetBrains.ReSharper.Plugins.FSharp.Psi.Parsing
+{
+  public class FSharpPreprocessedLexer : ILexer<FSharpLexerState>
+  {
+    private TokenNodeType currTokenType;
+    private readonly FSharpLexer myLexer;
+    private readonly FSharpPreprocessor myPreprocessor;
+    private readonly HashSet<string> myDefinedConstants;
+    private PreprocessorState myState = new PreprocessorState();
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi/src/Parsing/FSharpPreprocessedLexer.cs]: Lots of R# suggestions in file, like the one about making this field readonly. Were they ignored due to some reason?

**Comment by auduchinok** [comment]
Ah, I see. Cases were added to `dead code` cases in lexer tests.

**Pull request merged**